Internal Bookkeeping for MDC2020

AndrewEdmonds11 opened the pull request:
Hi Everyone

This is a draft pull request for adding internal bookkeeping of simulation stage efficiencies. Based on the feedback I received from the last Comp-Soft meeting, I have implemented this as using the ProditionsService. I am open to more comments and suggestions, including the names of the new classes I've added.

Here is an outline of the important features:
* I have added a new `SimStageEfficiency` data product, which we calculate using `SimStageEfficiencyCalculator_module` and store in the `art::Run`
    * a few example fcls have been added to `JobConfig/beam`
* I have added a printer for the new data product that can be used by the standard `PrintModule` and is also used by a new `PrintSimStageEfficiencies_module` to write out the efficiencies in text-file format for the ProditionsService (i.e. the text file produced can be used for testing and for adding to the real db)
    * added a fcl file that can be used like this: 
`mu2e -c Print/fcl/printSimStageEffsToDbTxt.fcl -S filelist-with-all-eff-files.txt`
    * here is the text file produced from my small-scale tests:
```
TABLE SimEfficiencies
# tag,numerator,denominator,efficiency
dsRegionPerPOT,661337,990000,0.668017
muBeamPerPOT,178708,990000,0.180513
flashPerMuBeam,159502,19900000,0.00801518
muStopPerPOT,97,70000,0.00138571
```
* the new `SimEfficiencies` DbTable does not require hard-coded efficiency tags
* the new DbTable is interfaced by the `Bookkkeeper` ProditionsEntitiy in new `SimulationConditions` and `SimulationConfig` packages
* `MixBackgroundFrames_module` gets the efficiencies (through the ProditionsService) that are given in the new fhicl parameter `simStageEfficiencyTags`
    * i.e. `simStageEfficienyTags : [ "muBeamPerPOT", "flashPerMuBeam" ]`

Let me know of any comments, questions, or suggestions you have.

Cheers
Andy

FNALbuild commented:
Hi @AndrewEdmonds11,
You have proposed changes to files in these packages:
- EventMixing
- ProditionsService
- JobConfig
- DbTables
- MCDataProducts
- Print
- EventGenerator
- SimulationConfig
- SimulationConditions

which require these tests: build.




<a href="https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild">FNALbuild is explained here.</a>

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 4b376e580a253dd36580ae4407a26c5189553c65: build

FNALbuild commented:
:sunny: The tests passed at ref 4b376e580a253dd36580ae4407a26c5189553c65.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 4b376e580a253dd36580ae4407a26c5189553c65 at 9de9f75aad5df33fa51d1b0196e785f73941a44a |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/scons.log). Build time: 12 min 12 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 16 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 6 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//183/console).
Build artefacts are deleted after 5 days. If this is not desired, select  on the job page.

kutschke commented:
If I understand correctly, the Bookkeeper object lives inside productions and it accumulates information as the job runs - did I get that right?

AndrewEdmonds11 commented:
Hi Rob. That's not quite right. The efficiencies aren't "accumulated" as a production stage runs, instead we run the `SimStageEfficiencyCalculator_module` after a stage has finished on all the output files for a given dataset. This module uses the `mu2e::GenEventCount` and `art::TriggerResults` data products to calculate the efficiency and creates a `SimStageEfficiency` object in the `art::Run`. 

The `Bookkeeper` stores all the efficiencies and is the object other modules use to access this information. The `Bookkeeper` is filled with information from the ProditionsService (either from fcl, the db itself or a db text file override).

Let me know if that explanation is unclear. I can also try to draw a diagram.

rlcee commented:
Rob and I just a few days ago had a discussion about whether Run products might preclude splitting files.  We can chose one or the other, but to accept both might take some art work.  I'm not sure, just highlighting the problem.

gaponenko commented on EventGenerator/src/SimStageEfficiencyCalculator_module.cc:
```diff
@@ -0,0 +1,111 @@
+      if (_conf.printTriggers()) {
+	std::cout << i_trigPath << std::endl;
+      }
+      if (i_trigPath == _conf.trigPath()) {
+	if (tnav.accepted(i_trigPath)) {
+	  ++_passedEvents;
+	}
```
Another example of white space issues:  the indentation of this block looks wrong in github;
because of hard tabs.   If you use emacs, try adding the block below to your $HOME/.emacs

; ----------------------------------------------------------------
;; See http://www.emacswiki.org/emacs/WhiteSpace
(require 'whitespace)
(setq whitespace-style '(face empty tabs trailing))
(add-hook 'c++-mode-hook
            '(lambda ()
                    (whitespace-mode)))
(add-hook 'c++-mode-hook
          '(lambda ()
             (setq indent-tabs-mode nil)))
(add-hook 'sh-mode-hook
            '(lambda ()
                    (whitespace-mode)))
(add-hook 'sh-mode-hook
          '(lambda ()
             (setq indent-tabs-mode nil)))
(add-hook 'latex-mode-hook
            '(lambda ()
                    (whitespace-mode)))
(add-hook 'latex-mode-hook
          '(lambda ()
             (setq indent-tabs-mode nil)))
;; ----------------------------------------------------------------

kutschke commented:
> The `Bookkeeper` stores all the efficiencies and is the object other modules use to access this information. The `Bookkeeper` is filled with information from the ProditionsService (either from fcl, the db itself or a db text file override).
> 
> Let me know if that explanation is unclear. I can also try to draw a diagram.

Thanks for clearing things up.  I was reacting to the presence of addEff and setEffVal member functions.  I now see that addEff is only used in the Maker class and setEffVal is not used anywhere.  Do you use setEffVal outside of this PR?

When I was doing the geometry and the orignal ConditionsService, I made the setters private and friended them to the Maker.  Or the friended Maker reached set the member data directly.   The idea was that public interface is explicitly const.   I know that the first few people to extend the geometry followed the same pattern but I have not checked to see how well this practice has been kept up.

I suggest that we adopt something similar for Proditions.

rlcee commented:
> > The `Bookkeeper` stores all the efficiencies and is the object other modules use to access this information. The `Bookkeeper` is filled with information from the ProditionsService (either from fcl, the db itself or a db text file override).
> > Let me know if that explanation is unclear. I can also try to draw a diagram.
> 
> Thanks for clearing things up. I was reacting to the presence of addEff and setEffVal member functions. I now see that addEff is only used in the Maker class and setEffVal is not used anywhere. Do you use setEffVal outside of this PR?
> 
> When I was doing the geometry and the orignal ConditionsService, I made the setters private and friended them to the Maker. Or the friended Maker reached set the member data directly. The idea was that public interface is explicitly const. I know that the first few people to extend the geometry followed the same pattern but I have not checked to see how well this practice has been kept up.
> 
> I suggest that we adopt something similar for Proditions.

DbHandle and ProditionsHandle only return const refs or pointers.  Is this sufficient to address your concern?

rlcee commented:
Based on my admittedly shaky memory of Andy's talk, we needed to have a discussion of how run numbers are to be used in simulation.  I think a few points came up.

1. are run numbers only used to represent actual time dependence of the data (as opposed to different background samples, for ex.)
2. will simulation of time dependence of the data be done using real data run numbers or real number+plus 1M, or a representative scheme, such as one fake run number per data run period?
3.  will the mixing rates table be run dependent? representing what?
4. what run numbers should we be using now, for what purposes?

I created a [run number page](https://mu2ewiki.fnal.gov/wiki/RunNumbers) in the wiki.  I don't think we need to commit to decisions now for all time, but I do think we need a working agreement to verify the code is consistent with that agreement.

rlcee commented:
Assuming the mixing rate table is static (not run-dependent), another significant issue is the use of the database w.r.t. these quantities.  In the process of developing the conditions database, we discussed static entries such as geometry, particle lists, and module configurations.  At that time, we said that the code we were developing would not be designed for these items.  Maybe now we need a new policy on static quantities.  Here are some options

1. disallow static quantities in the condition database, leave in fcl and file-based conditions
2. shoehorn non-run-dependent quantities into the run dependent structure, creating unnecessary maintenance
3. design a proper non-run-dependent database system.  This might be much simpler than the run-dependent in maintenance and access.  We would also have to design how to specify the selection of quantities (tags), which might be quite broad and varied, depending on the policy.

It doesn't seem obvious what is the most flexible, reproducible, robust, user-friendly, and low-maintenance.  In my opinion, to first order, moving a number from fcl to the database introduces considerable maintenance, infrastructure dependence/fragility, and a run-time delay.

gaponenko commented:
Hi,

On Mon, 24 Aug 2020, Ray Culbertson wrote:

> Rob and I just a few days ago had a discussion about whether Run products might preclude splitting files.  We can chose one or the other, but to accept both might take some art work.  I'm not sure, just highlighting the problem.  
> 

I like the idea to use Conditions for efficiency bookkeeping because
it makes the Run product unnecessary.  I think the code can be cleaned
up to not put anything in Run.  One would process a dataset reading
events (for trigPath result) and SubRuns (for GenEventCount), then
write out the aggregated information directly from that job and enter
it into the database.

Is the data format for database entry the same as what one needs for a
a FHICL override (DB bypass)?  If not, then write out both formats
from that job to two different files.

Andrei

> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_234-23issuecomment-2D679408829&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=XfcCtT4GHU9pvtU4crFeGuvljZVMuQElo1X8FgGmr74&s=NkugOLUL-8ejcOo2MaRz7QUoRCfWQB-iG8ywl1bk1Bk&e=

gaponenko commented:
Hi Ray,

I do not understand the complication from "no run dependency".  I am
under the impression that the system that you designed can already
handle this use case.  The run validity range can be set to cover all
runs in a given MC production, and the version can be set to something
like MDC2018 or MCD2020 etc., to match the efficiency values to
the corresponding simulation.   What am I missing?

Andrei

On Tue, 25 Aug 2020, Ray Culbertson wrote:

> Assuming the mixing rate table is static (not run-dependent), another significant issue is the use of the database w.r.t. these quantities.  In the process of developing the conditions database, we discussed static entries such as geometry, particle lists, and module configurations.  At that time, we said that the code we were developing would not be designed for these items.  Maybe now we need a new policy on static quantities.  Here are some options
> 
> 1. disallow static quantities in the condition database, leave in fcl and file-based conditions
> 2. shoehorn non-run-dependent quantities into the run dependent structure, creating unnecessary maintenance
> 3. design a proper non-run-dependent database system.  This might be much simpler than the run-dependent in maintenance and access.  We would also have to design how to specify the selection of quantities (tags), which might be quite broad and varied, depending on the policy.
> 
> It doesn't seem obvious what is the most flexible, reproducible, robust, user-friendly, and low-maintenance.  In my opinion, to first order, moving a number from fcl to the database introduces considerable maintenance, infrastructure dependence/fragility, and a run-time delay.
> 
> 
> 
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_234-23issuecomment-2D680129465&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=f8_1RTiC-Ac3I7QcxNWv3OCW8LvccxdymKdJh2gkXf8&s=55jgOv-a5xRGXzUILNGmjdQMkHjAxz2jMlMlAV9LVDI&e=

rlcee commented:
> Hi,
> On Mon, 24 Aug 2020, Ray Culbertson wrote: Rob and I just a few days ago had a discussion about whether Run products might preclude splitting files. We can chose one or the other, but to accept both might take some art work. I'm not sure, just highlighting the problem.
> I like the idea to use Conditions for efficiency bookkeeping because it makes the Run product unnecessary. I think the code can be cleaned up to not put anything in Run. One would process a dataset reading events (for trigPath result) and SubRuns (for GenEventCount), then write out the aggregated information directly from that job and enter it into the database. Is the data format for database entry the same as what one needs for a a FHICL override (DB bypass)? If not, then write out both formats from that job to two different files. Andrei
> […](#)
> -- You are receiving this because your review was requested. Reply to this email directly or view it on GitHub: https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_234-23issuecomment-2D679408829&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=XfcCtT4GHU9pvtU4crFeGuvljZVMuQElo1X8FgGmr74&s=NkugOLUL-8ejcOo2MaRz7QUoRCfWQB-iG8ywl1bk1Bk&e=

The format of quantities in fcl is not directly tied to the format from the database (Prodition entity fromFcl method can be different from fromDb), but one would expect them to be recognizably one-to-one.  When the contents of a database table are represented in a text file, the correspondence is exact.

The system allows tables to be committed from a local (not grid) art end-job but, in my opinion, I would not blind-fill tables unless the load is beyond human scale and very simple/reliable at the same time.  You might have forgotten your kerberos principle, and the commit fails, for example.  Or just a sanity check.

If the mixing table contains 10 values, each of which is to be written by a different art exe end-job, then they can't be written to one table.  A table must be committed completely in one commit.  Filling one table with multiple commands would require a significant re-write.  One could imagine 10 tables, but that's not appealing either.  I always imagined the table contents gathered in text, then one commit with dbTool.

The idea that a dataset has a derived quantity doesn't fit in SAM or conditions models, as far as I can see.  It might be useful to develop a dataset quantity system.  In SAM, you can store arbitrary values, but you have to write them for files, not datasets.

rlcee commented:
> Hi Ray, I do not understand the complication from "no run dependency". I am under the impression that the system that you designed can already handle this use case. The run validity range can be set to cover all runs in a given MC production, and the version can be set to something like MDC2018 or MCD2020 etc., to match the efficiency values to the corresponding simulation. What am I missing? Andrei
> […](#)
> On Tue, 25 Aug 2020, Ray Culbertson wrote: Assuming the mixing rate table is static (not run-dependent), another significant issue is the use of the database w.r.t. these quantities. In the process of developing the conditions database, we discussed static entries such as geometry, particle lists, and module configurations. At that time, we said that the code we were developing would not be designed for these items. Maybe now we need a new policy on static quantities. Here are some options 1. disallow static quantities in the condition database, leave in fcl and file-based conditions 2. shoehorn non-run-dependent quantities into the run dependent structure, creating unnecessary maintenance 3. design a proper non-run-dependent database system. This might be much simpler than the run-dependent in maintenance and access. We would also have to design how to specify the selection of quantities (tags), which might be quite broad and varied, depending on the policy. It doesn't seem obvious what is the most flexible, reproducible, robust, user-friendly, and low-maintenance. In my opinion, to first order, moving a number from fcl to the database introduces considerable maintenance, infrastructure dependence/fragility, and a run-time delay. -- You are receiving this because your review was requested. Reply to this email directly or view it on GitHub: https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_234-23issuecomment-2D680129465&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=f8_1RTiC-Ac3I7QcxNWv3OCW8LvccxdymKdJh2gkXf8&s=55jgOv-a5xRGXzUILNGmjdQMkHjAxz2jMlMlAV9LVDI&e=

 Yes, you can do this -  it is what I meant by "shoehorn non-run-dependent quantities into the run dependent structure, creating unnecessary maintenance"

AndrewEdmonds11 commented:
Thanks for the comments, Andrei :)

>     * In DbTable::Row,  do not persist _eff.  Recompute it on use just like
>       is already done in SimStageEfficiency.hh
If I do this, then the efficiency won't appear in the database/text file as a single number, which I think is preferable because then it would be easier to check.

>     * Be explicit with "override" every time a virtual function from a
>       base class is overridden.  I think SimEfficiencies should have some
>       overrides.
Will do.

>     * Only a single instance of GenEventCount should exist in subrun.
>       There is no need for genEventCountTag config.  See
>       Analyses/src/GenEventCountReader_module.cc Without this check we'd
>       have datasets with multiple GenEventCount objects added by mistake
>       at different simulation stages.
Good point. I can remove that fhicl parameter and get only the GenEventCount that exists in the file.

>     * Is the name "Bookkeeper" too generic?
>       SimulationConditions/inc/Bookkeeper.hh
Yes, I agree. Maybe I could rename to `SimBookkeeper`?

>     * There are whitespace errors in some files.  Many files lack the
>       newline at the end.  Look for github mark up, e.g. at the end of
>       JobConfig/beam/DS-resampler-TrkCalCRVEff.fcl
Thanks for catching those. I will add the .emacs code you suggested inline and fix the whitespace.

>     * Can you please explain the reason to switch between branchName()
>       and moduleLabel()  in SimStageEfficiencyPrinter.cc ?
I have set-up SimStageEfficiencyPrinter to have two modes. One for writing to the Db text file and one to match what is done in the other printers. In the normal printers we print the full branchName, and for the Db text file, we only require the module label.
 
>     * Can we replace the two functions
>       Print(const art::Handle& handle, ...)
>       Print(const art::ValidHandle& handle,...)
>       with one
>       Print(const SimStageEfficiency& eff, ...)?
>       (By dereferencing the handles at function call, not inside its
>       implementation.)
This is inherited from `ProductPrinter` so Ray might want to comment on this.

AndrewEdmonds11 commented:
Hi Rob
> Thanks for clearing things up. I was reacting to the presence of addEff and setEffVal member functions. I now see that addEff is only used in the Maker class and setEffVal is not used anywhere. Do you use setEffVal outside of this PR?

No, I don't sorry. This is left over code that I will remove...

AndrewEdmonds11 commented:
Thanks for the discussion everyone. I might have some thoughtful comments later but to clarify one thing:

> If the mixing table contains 10 values, each of which is to be written by a different art exe end-job, then they can't be written to one table. 

The way this PR is currently set up, all 10 values would be written to the table in one job. Although each `SimStageEfficiency` data product is in a different file. The idea is that one would run `PrintSimStageEfficiencies_module` on a filelist that contains all the files. I don't think that changes any of the underlying points though...

kutschke commented:
Will everyone be at the Computing and Software meeting tomorrow?

If so, I would like everyone to come prepared to discuss the DB issue that has been raised here: is the conditions db the right place for this information?  If not what is the right place and what other sorts of information will live in that place?

There is a related issue that we started to address, and never finished, during one of Richie's PRs that was eventually closed: allocating ranges of run numbers for different uses; one use is for simulations using nominal values of quantities that will later need detailed calibration.   Ideally I would like to resolve this now.  It may be good enough to identify a things we need to decide now and things we can defer - then decide the subset that is needed now.

brownd1978 commented:
Hi Ray,

On Tue, Aug 25, 2020 at 9:23 AM Ray Culbertson <notifications@github.com>
wrote:

> Assuming the mixing rate table is static (not run-dependent), another
> significant issue is the use of the database w.r.t. these quantities. In
> the process of developing the conditions database, we discussed static
> entries such as geometry, particle lists, and module configurations. At
> that time, we said that the code we were developing would not be designed
> for these items. Maybe now we need a new policy on static quantities. Here
> are some options
>
The mixing rates (efficiencies) can be run dependent, for instance if we
simulate different detector setups (complete/partial tracker, ...) that
correspond to different real data taking periods.   In
general, our database design must accommodate run dependent simulations
conditions, efficiencies are just 1 example.


>    1. disallow static quantities in the condition database, leave in fcl
>    and file-based conditions
>
> As a general principle I agree with this.  But I don't think it applies to
these efficiencies as stated above.


>
>    1. shoehorn non-run-dependent quantities into the run dependent
>    structure, creating unnecessary maintenance
>
> We need a MC run numbering scheme that allows us to track real data run
numbers, as you pointed out earlier.  As we have no real data right now,
any scheme we use for MDC2020 will be artificial.  Is that what you mean
by shoehorn?
I don't see how this creates unnecessary maintenance, as this is a
necessary feature.


>    1. design a proper non-run-dependent database system. This might be
>    much simpler than the run-dependent in maintenance and access. We would
>    also have to design how to specify the selection of quantities (tags),
>    which might be quite broad and varied, depending on the policy.
>
> It doesn't seem obvious what is the most flexible, reproducible, robust,
> user-friendly, and low-maintenance. In my opinion, to first order, moving a
> number from fcl to the database introduces considerable maintenance,
> infrastructure dependence/fragility, and a run-time delay
>
.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/234#issuecomment-680129465>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575ZAQQN2WE4HKENECTSCPQQNANCNFSM4QJZEARQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
Hi Andy,
A key issue in your updated design is translating from art-format event
efficiencies to database entries.  Since there are issues associating the
event count data with Runs, and since all sim subruns are unique, I suggest
you move your data product to subrun.  I suggest production write these
products to a dedicated stream that has no per-event content (in addition
to storing it in the event payload stream) directly to DCache disk
(persistent).
These tiny event count data product files could be quickly converted to
database efficiency tables by a standalone process (still part of the POMS
campaign) that wouldn't require any prestaging or be subject to
DCache flushing of on-disk data.  These are essentially fhicl changes to
the design in your PR, but I think it helps to think through the whole
process to see where the code design might need work.
Dave

On Tue, Aug 25, 2020 at 10:29 AM Andrew Edmonds <notifications@github.com>
wrote:

> Thanks for the discussion everyone. I might have some thoughtful comments
> later but to clarify one thing:
>
> If the mixing table contains 10 values, each of which is to be written by
> a different art exe end-job, then they can't be written to one table.
>
> The way this PR is currently set up, all 10 values would be written to the
> table in one job. Although each SimStageEfficiency data product is in a
> different file. The idea is that one would run
> PrintSimStageEfficiencies_module on a filelist that contains all the
> files. I don't think that changes any of the underlying points though...
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/234#issuecomment-680166643>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH5774WTECFIOYLR7ZEGTSCPYITANCNFSM4QJZEARQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
> >     * There are whitespace errors in some files.  Many files lack the
> >       newline at the end.  Look for github mark up, e.g. at the end of
> >       JobConfig/beam/DS-resampler-TrkCalCRVEff.fcl
> Thanks for catching those. I will add the .emacs code you suggested inline and fix the whitespace.

One clarification: the emacs code I posted does not fix any whitespace
automatically.  It only highlights hard tabs and trailing whitespace.
One will still need to run "M-x untabify" and 
"M-x delete-trailing-whitespace" by hand.

To ensure each text files ends with a newline automatically there is this:

(setq require-final-newline t)

gaponenko commented:
Hi,

> >     * In DbTable::Row,  do not persist _eff.  Recompute it on use just like
> >       is already done in SimStageEfficiency.hh
> If I do this, then the efficiency won't appear in the database/text file as a single number, which I think is preferable because then it would be easier to check.

What I find objectionable is storing redundant information that can
potentially end up in a self-inconsistent state.   Something like

numerator=50 demoninator=100 eff=0.1

Even if your code would never write this, the reading code would
either ignore a piece of data (confusing to someone looking at the DB:
did the reader ignore 50 or 0.1?)  or check that the record is
self-consistent (extra work, not trivial with floating point numbers).

If you want to store the derived quantity, a way to do it without
information loss is to store just  denominator and eff.

Andrei

gaponenko commented:
> >     * Can we replace the two functions
> >       Print(const art::Handle& handle, ...)
> >       Print(const art::ValidHandle& handle,...)
> >       with one
> >       Print(const SimStageEfficiency& eff, ...)?
> >       (By dereferencing the handles at function call, not inside its
> >       implementation.)
> This is inherited from `ProductPrinter` so Ray might want to comment on this.

Reducing the interface to a single Print() looks attractive to me, but is outside
of the scope of this PR.   For the moment I suggest to have each of the two
required Print(Handle& h), Print(ValidHandle& h) function to be implemented like

Print(Handle<SimStageEfficiency>& h) { Print(*h); }

with Print(SimStageEfficiency&) implementing the actual logic, instead
of duplicating the logic as is the case now.

Andrei

gaponenko commented:
Hi,

On Tue, 25 Aug 2020, Ray Culbertson wrote:

> 
> The system allows tables to be committed from a local (not grid) art end-job but, in my opinion, I would not blind-fill tables unless the load is beyond human scale and very simple/reliable at the same time.   You might have forgotten your kerberos principle, and the commit fails, for example.  Or just a sanity check.

Agreed.  By the "write" in my message I meant to write into a text
file, which is inspected before loading it into the database.

> 
> If the mixing table contains 10 values, each of which is to be written by a different art exe end-job, then they can't be written to one table.  A table must be committed completely in one commit.  Filling one table with multiple commands would require a significant re-write.  One could imagine 10 tables, but that's not appealing either.  

Having multiple tables looks like the correct granularity for me.
They are effieciencies for different simulation jobs.  One may need to
re-run a stage and update e.g.  CRV hit efficiency for neutrons
without wanting to change the number of stopped muons per POT or
number of DIO tracker hits per muon.

Andrei

gaponenko commented on SimulationConditions/inc/BookkeeperMaker.hh:
```diff
@@ -0,0 +1,43 @@
+    typename Bookkeeper::ptr_t fromDb(typename SimEfficiencies::cptr_t effDb) {
+      // fill the Bookkeeper with initial values
+      auto ptr = fromFcl();
+      // now overwrite with values from database
```
I do not see a use case for providing a value in FCL and then overriding it with a DB record, as the present code does.  It would make more sense to load DB info first, and allow to override a DB value from FCL.

rlcee replied:
> I do not see a use case for providing a value in FCL and then overriding it with a DB record, as the present code does. It would make more sense to load DB info first, and allow to override a DB value from FCL.

 This style was in the context of writing all your proditions code first as filled from fcl.  At a later step, you start overriding  pieces from the database as you develop the database context.  During this process, you have the option to turn off the database at any time and get good fcl defaults.  If you were in the situation where well-developed and reading from the database regularly, you can supply temporary overrides of the database result by the text file method.

But this raises another issue that was not fully discussed.  Currently Proditions in TrackerConditions comes 90% from fcl and 10% (potentially) from the database.  Should there be a rule or assumption that all Proditions inputs will eventually come the database or does it make sense to allow some to remain in fcl?  Are some parameters more like reco configuration rather than describing the data?

brownd1978 replied:
We are working towards a majority of tracker proditions coming from the database for MDC2020.  The distinction between configuration (for simulation) and conditions (for reconstruction) is blurry.  If we imagine simulating different run periods with different configurations, then all the 'configurations' need to be conditions, so that they can be treated coherently in sim and reco.

rlcee commented:
> Hi Ray,
> On Tue, Aug 25, 2020 at 9:23 AM Ray Culbertson ***@***.***> wrote: Assuming the mixing rate table is static (not run-dependent), another significant issue is the use of the database w.r.t. these quantities. In the process of developing the conditions database, we discussed static entries such as geometry, particle lists, and module configurations. At that time, we said that the code we were developing would not be designed for these items. Maybe now we need a new policy on static quantities. Here are some options
> The mixing rates (efficiencies) can be run dependent, for instance if we simulate different detector setups (complete/partial tracker, ...) that correspond to different real data taking periods. In general, our database design must accommodate run dependent simulations conditions, efficiencies are just 1 example.
> 1. disallow static quantities in the condition database, leave in fcl and file-based conditions As a general principle I agree with this. But I don't think it applies to
> these efficiencies as stated above.
> 1. shoehorn non-run-dependent quantities into the run dependent structure, creating unnecessary maintenance We need a MC run numbering scheme that allows us to track real data run
> numbers, as you pointed out earlier. As we have no real data right now, any scheme we use for MDC2020 will be artificial. Is that what you mean by shoehorn? I don't see how this creates unnecessary maintenance, as this is a necessary feature.
> 1. design a proper non-run-dependent database system. This might be much simpler than the run-dependent in maintenance and access. We would also have to design how to specify the selection of quantities (tags), which might be quite broad and varied, depending on the policy. It doesn't seem obvious what is the most flexible, reproducible, robust, user-friendly, and low-maintenance. In my opinion, to first order, moving a number from fcl to the database introduces considerable maintenance, infrastructure dependence/fragility, and a run-time delay
> .
> […](#)
> — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#234 (comment)](https://github.com/Mu2e/Offline/pull/234#issuecomment-680129465)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABAH575ZAQQN2WE4HKENECTSCPQQNANCNFSM4QJZEARQ> .
> -- David Nathan Brown Dave_Brown@lbl.gov Office Phone (510) 486-7261 Fax 495-2957 Lawrence Berkeley National Lab MS 50R5008 (50-6026C) Berkeley, CA 94720

You pointed out a case I was not considering - when you have run dependence in the conditions for the simulation, that will be reflected in the efficiency of the skims, which will in turn affect the mix-in rate.  This is a convincing case that there is run dependence in these numbers.

In the case of run-dependent mixing, how do you make sure the right runs or subruns are available in the mixing job?  When you are producing mixed run X, you only provide mix-in files with run X?

Andrei and you are arguing that using the database eliminates the Run or SubRun products at least at the later stages.  I'm not  convinced which is easier or more fool-proof.  

Would it be enough for each skimmed event to know how many events were rejected since the last accepted event?

gaponenko commented:
Hi,

On Tue, 25 Aug 2020, Ray Culbertson wrote:

> Andrei and you are arguing that using the database eliminates the Run or SubRun products at least at the later stages.  I'm not  convinced which is easier or more fool-proof.  

Having the numbers in FCL was OK with me; putting them in conditions
is also fine.  Using Run products from separate (from event data)
files for this bookeeping feels a lot messier to me.  Very contrived.
I'd rather write out the numbers in the fcl format and include that
fcl in the main job config, so that the cut-and-paste error argument
goes away.

> Would it be enough for each skimmed event to know how many events were rejected since the last accepted event? 

No.  We'd lose information from jobs with no events passed.  Also in
each job, lose events after the last accepted event.  I think we
discussed this before.

Andrei

brownd1978 commented:
Using art output stream to save intermediate event counts (efficiencies) from pre-mixing production allows using existing production infrastructure to store the results.  If we write fcl or other ad-hoc formats we will need to create new technology to transfer and combine and bookkeep those.  I agree there are open questions how these intermediate collections are processed and turned into proditions objects in the database, and how mixing jobs should find the correct efficiencies to associate with a given mixin stream.

gaponenko commented on EventMixing/src/MixBackgroundFrames_module.cc:
```diff
@@ -92,6 +98,16 @@ namespace mu2e {
+          std::vector<std::string>()
+          };
+
+      fhicl::Sequence<double> meanEventsPerPOTFactors{ Name("meanEventsPerPOTFactors"),
```
Are meanEventsPerPOTFactors some fudge factors that should approximately never be used?   It should not be a required piece of configuration, make it an OptionalSequence.

kutschke commented:
Hi Andy,

     I am reviewing this thread before the meeting today.   Your diagrams in https://mu2e-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=34959 have been a big help.

     I have a few workflow mechanics related questions.  I don't think that impact the big question but I would feel more comfortable if I am sure that I have the right mental model of the big picture.

     On page 2 of the diagram.  

I think that the mubeam-eff output files and TrkCalCRVEff output files are regular art files.  Is that right?
Do you envisage a one-one map of these files to the files that are outputs of stage 1?  1:many?
Are these small files that could reasonably be disk resident (in persistent) for the duration of a campaign?  ( with just run and subrun products; or do they also contain event scope products?)

I also understnd that PrintSimStageEfficiencies is a single job that reads everything from the previous stage ( both beam and resampler ) and writes it to a temporary file that will go to the "official place" ( which you propose to be Proditions ).  Is that the right picture.

Thanks

AndrewEdmonds11 commented:
Hi Rob

> I think that the mubeam-eff output files and TrkCalCRVEff output files are regular art files. Is that right?

That's correct. These are regular art files

> Do you envisage a one-one map of these files to the files that are outputs of stage 1? 1:many?

I envisage it being many:1. We would run over all the outputs of stage 1 in a single job and create a single art file with the SimStageEfficiency

> Are these small files that could reasonably be disk resident (in persistent) for the duration of a campaign? ( with just run and subrun products; or do they also contain event scope products?)

On my test job of 990k POT the mubeamEff file is 3.9 MB, and the dsregionEff file is 13MB. I'm not sure how this will scale with statistics. There could be some optimization to do here... They just contain the run product so I'm surprised the files are a few MB,,,

> I also understnd that PrintSimStageEfficiencies is a single job that reads everything from the previous stage ( both beam and resampler ) and writes it to a temporary file that will go to the "official place" ( which you propose to be Proditions ). Is that the right picture.

Yes, that's correct.

kutschke commented:
> > Do you envisage a one-one map of these files to the files that are outputs of stage 1? 1:many?
> 
> I envisage it being many:1. We would run over all the outputs of stage 1 in a single job and create a single art file with the SimStageEfficiency

On a long simulation campaign, the files that are output from the early parts of stage 1 might disappear from cache before the last files are finished.   There are lots of ways to redesign this part of the process to avoid having to stage in files just to read these data products.  No need to solve it today.


> > Are these small files that could reasonably be disk resident (in persistent) for the duration of a campaign? ( with just run and subrun products; or do they also contain event scope products?)
> 
> On my test job of 990k POT the mubeamEff file is 3.9 MB, and the dsregionEff file is 13MB. I'm not sure how this will scale with statistics. There could be some optimization to do here... They just contain the run product so I'm surprised the files are a few MB,,,
>

Even scaled up this fits comfortably in dCache.

kutschke commented:
@FNALbuild test

ryuwd commented:
I don't see any issues with our CI. I suspect something has gone wrong with the webhook or Jenkins. Patrick may need to investigate.

The job history doesn't show any webhooks (comments, git pushes, etc) being received from our repository recently, or any repo at all today : https://buildmaster.fnal.gov/buildmaster/view/GitHub%20Integration/job/GithubIntegration/job/dispatch-github-webhook/

Best wishes, Ryun

ryuwd commented:
comment to trigger jenkins

FNALbuild commented:
:hourglass: The following tests have been triggered for 1813047d983b04521cafef9e783655d4b175170e: build

ryuwd commented:
also a code checks job, which might help rectify any whitespace issues

@FNALbuild run code checks

(manually triggered - webhook is still broken)

FNALbuild commented:
:hourglass: The following tests have been triggered for 1813047d983b04521cafef9e783655d4b175170e: code checks

FNALbuild commented:
:x: Trailing whitespace characters were found at 1813047d983b04521cafef9e783655d4b175170e on files you changed.

You can review the generated patch [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//107/artifact/codechecks-pr234-1813047d983b04521cafef9e783655d4b175170e.patch).

If it is convenient, you may remove the trailing whitespace like this:
```
curl https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//107/artifact/codechecks-pr234-1813047d983b04521cafef9e783655d4b175170e.patch | git apply -v --index
git commit -am "Remove trailing whitespace found at 1813047d983b04521cafef9e783655d4b175170e" && git push
```

FNALbuild commented:
:sunny: The tests passed at ref 1813047d983b04521cafef9e783655d4b175170e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 1813047d983b04521cafef9e783655d4b175170e at 9de9f75aad5df33fa51d1b0196e785f73941a44a |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/scons.log). Build time: 13 min 41 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 9 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//187/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

brownd1978 commented:
@FNALbuild test

FNALbuild commented:
:hourglass: The following tests have been triggered for 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0: build  (Build queue has 2 jobs)

FNALbuild commented:
:sunny: The tests passed at ref 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0 at 9de9f75aad5df33fa51d1b0196e785f73941a44a |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/scons.log). Build time: 12 min 23 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 9 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//195/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

AndrewEdmonds11 commented:
Here's the final run-down of this PR before I close it and merge it in with PRs #231 and #243:

- added Proditions infrastructure to handle SimStageEfficiencies
- modified MixBackgroundFrames_module to use SimStageEfficiencies (and any other factors given)
- this is not a breaking change: previous behaviour can be recovered by using the old `meanEventsPerProton` fcl parameter, which overrides the new behaviour

The piece needed for internal bookkeeping is calculating the sim stage efficiencies. We decided to use a new script (outside of Offline) that will get the information from SAM.

ryuwd commented:
.

FNALbuild commented:
:memo: The HEAD of `master` has changed to ef8ebbd15963a2b5d2f2f8d51a3760824713eaa1. Tests are now out of date.

FNALbuild commented:
:memo: The HEAD of `master` has changed to ef8ebbd15963a2b5d2f2f8d51a3760824713eaa1. Tests are now out of date.

FNALbuild commented:
:memo: The HEAD of `master` has changed to ef8ebbd15963a2b5d2f2f8d51a3760824713eaa1. Tests are now out of date.

ryuwd commented:
Apologies for the spam.

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0: build  (Build queue has 2 jobs)

FNALbuild commented:
:sunny: The tests passed at ref 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 7b3d72a36de7f3767be5d8bbb1cf4d0471ea73f0 at ef8ebbd15963a2b5d2f2f8d51a3760824713eaa1 |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/scons.log). Build time: 12 min 41 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (0) FIXME (1) in 9 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [10 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//229/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/56fc3a161da40ff8ba367e05ce2bc727)
